c10/core/TensorImpl.h

```diff
@@ -1184,6 +1178,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
         }
       }
       data_type_ = meta;
+      // NB: device is not changed
```

Comment:
I'd rather not do https://github.com/pytorch/pytorch/pull/18605 and instead reset the device - around line 1215 below. Non-initialized tensors in C2 can be created at random places (e.g. in static fields) but the actual data pointer would be populated only at raw_mutable_data() call and the device may have changed in the mean time. It's a weird behavior, but still...